proton

```diff
@@ -28,10 +28,18 @@ from ctypes import c_void_p
 from filelock import FileLock
 from random import randrange
 
+try:
+    # available since 3.3
+    from shlex import quote as shell_quote
+except ImportError:
+    # exists in older versions, undocumented
+    from pipes import quote as shell_quote
+
 #To enable debug logging, copy "user_settings.sample.py" to "user_settings.py"
 #and edit it if needed.
 
 CURRENT_PREFIX_VERSION="7.0-100"
+PROTON_APP_ID="1887720"
```

Comment:
I believe the intended design on the Steam side is that if a developer has requested that the game is run with the ability to inject arbitrary debugging commands like this, then the Steam client will iterate through the stack of compatibility tools and identify whichever one is the last one before the game itself:

```
Steam client -> SteamLinuxRuntime_soldier -> Proton (whatever version) -> a Windows game
                                                      ^^^

Steam client -> SteamLinuxRuntime_soldier -> SteamLinuxRuntime -> an old native Linux game
                                                      ^^^

Steam client -> SteamLinuxRuntime_sniper -> a future native Linux game
                         ^^^
```

and then set `STEAM_COMPAT_LAUNCHER_SERVICE` to the app-ID of that tool, so that the arbitrary commands get injected into something that is as close as possible to the execution environment of the actual game.

If compatibility tools have a unique identifier, then we could use that unique identifier instead of the app-ID (assuming @ttimo or his colleagues can implement that on the Steam side); but I don't have visibility into the `compatibilitytool.vdf` of the various SteamLinuxRuntime tools, so I don't know what their unique identifiers are. I do know what their app IDs are, and their app IDs are how Proton's `toolmanifest.vdf` refers to them in any case, so that seemed a reasonable unique identifier to work with.

Strictly speaking, it only needs to be unique within a "stack" of compatibility tools: tools A and B only need to be distinguished if it would make sense to run tool A "inside" tool B or vice versa. But it needs to be something that the Steam client knows, so using an existing metadata field seemed easier than inventing a new one.

> how is this supposed to work with tools that are dropped in `compatibilitytools.d/`?

It won't, because Steam can't know an app-ID for those. But if the developer sets `STEAM_COMPAT_LAUNCHER_SERVICE=proton`, then Steam will ignore it and pass it through, and then the startup script will recognise that as a request to have Proton be the layer that starts the `launcher-service`.

> Also proton being alias to 1887720 feels wrong. Will you point it to Proton 8 (or whatever version the next rabase will be based on) when it releases? Maybe it should be just proton7?

There's no need to distinguish between `proton7` and `proton8` unless there would be a legitimate scenario in which the stack of compatibility tools includes more than one version of Proton (Proton running inside Proton), which I don't think would ever make sense? So the alias that's there for developer convenience is just `proton`, meaning "whatever stack of compatibility tools I'm using, if it includes some version of Proton, then I want the `launcher-service` to be inserted at that layer".

The app-ID here needs to match up with what the Steam client will set as the value of `STEAM_COMPAT_LAUNCHER_SERVICE`, which will be different for each Proton app-ID in the Steam catalogue. Proton has separate git branches for 5.13, 6.3, 7.0 and experimental, so I had assumed that it would be easy to change the app-ID in `./proton` when switching between git branches.

Alternatively, if it needs to be possible to set an app-ID "from the outside", then it could be injected via the build system instead, similar to the way the `$(build_name)` is handled: that's what I did for `SteamLinuxRuntime_soldier` and `SteamLinuxRuntime_sniper`, which are built from the same source code and just have different parameters passed in.

Reply:
This is no longer needed in today's revision.